src/QuantumAcademy.jl

```diff
@@ -48,15 +48,15 @@ function Base.getindex(H::FinDiffHamiltonian{T}, i::Int, j::Int) where T
     if ix == jx
         if iy == jy
             Vij = H.potential(T((ix-1) // H.nlength), T((iy-1) // H.nwidth))
-            return (T(H.nlength) / H.length)^2 + (T(H.nwidth) / H.width)^2 + π^2/2*(1/H.length^2*H.kpointx^2 + 1/H.width^2*H.kpointy^2) + Vij
+            return (Complex(H.nlength) / H.length)^2 + (Complex(H.nwidth) / H.width)^2 + π^2/2*(1/H.length^2*H.kpointx^2 + 1/H.width^2*H.kpointy^2) + Vij
         elseif abs(iy - jy) == 1 || abs(iy - jy) == H.nwidth - 1
-            return -(T(H.nwidth) / H.width)^2 / 2 - H.kpointy*H.nwidth/H.width^2*π/2*sign(iy-jy - fld(iy,H.nwidth)*H.nwidth + fld(jy,H.nwidth)*H.nwidth)im
+            return -(Complex(H.nwidth) / H.width)^2 / 2 - H.kpointy*H.nwidth/H.width^2*π/2*sign(iy-jy - fld(iy,H.nwidth)*H.nwidth + fld(jy,H.nwidth)*H.nwidth)im
         end
     elseif iy == jy && (abs(ix - jx) == 1 || abs(ix - jx) == H.nlength - 1)
-        return -(T(H.nlength) / H.length)^2 / 2 - H.kpointx*H.nlength/H.length^2*π/2*sign(ix-jx - fld(ix,H.nlength)*H.nlength + fld(jx,H.nlength)*H.nlength)im
+        return -(Complex(H.nlength) / H.length)^2 / 2 - H.kpointx*H.nlength/H.length^2*π/2*sign(ix-jx - fld(ix,H.nlength)*H.nlength + fld(jx,H.nlength)*H.nlength)im
     end
 
-    return zero(T)
+    return zero(Complex)
```

Comment:
zero(T) is wrong, since T is the type of the length/width parameters - that is actually an issue with the way FinDiffHamiltonian is implemented, since the implicit assumption is that the type of the elements of the Hamiltonian is the same as that in which the length is specified, which is not generally a reasonable thing to assume. Would probably be better in general to just have the user specify which type the elements should be, and assume ComplexF64 by default?

Reply:
The idea is to output `zero(T)` but have `getindex` convert to `Complex` of that implicitly by putting a type declaration on the output. I agree that my implementation had only real components and for your Bloch case you need to change the element type to `Complex{T} where T<:Real` rather than `T<:Real`.